### ISSUE TITLE
Try to solve =org-indent-mode= not compatible with =org-transclusion=

### ISSUE THREAD

**Issue by randomwangran** [nobiot/org-transclusion#89]
I see the known limitation on the Readme file about this issue, but I would like to have a try to push the boundary.

The following figure illustrates what the issue is. Two black boxes are two states to test `org-indent-mode` working with `org-transclusion`. The first one is without `org-indent-mode`. The indentation in the transclude headline: `this is a headline for testing` is properly indent. However, with `org-indent-mode` on, the headline, which is marked by the red arrow, is not indent properly. The red line is drawn to help illustration of the wrong indentation.

I am writing this thing first to confirm with the author that I correctly understand the known limitation written in the readme file.

![image](https://user-images.githubusercontent.com/8197831/127910497-6f4b82fa-971c-404c-abfa-66a2977a2511.png)

**Comment by nobiot** [comment]
@randomwangran 
I think your understanding is correct; it's a little difficult to see even with the comparison image...

The technical reason for this incompatibility is simple. Both modes use `line-prefix` and `wrap-prefix` text properties to control what's displayed on the UI. Org-indent uses them to create visual indentation (the actual text data in the filesystem and buffer don't have the indentation). Org-transclusion use them to add the vertical line in the fringe in GUI version of Emacs.

It might make the issue clear if I showed you the Terminal UI when both modes are turned on. See the image below from my Windows Terminal (it should be the same if you used any terminal emulator in your OS). As Emacs on a terminal cannot use the fringe, Org-trnasclusion falls back to adding a vertical line character, retaining the indentation (but the number of indentation should be reduced by -1 as the vertical bar adds one character).

![image](https://user-images.githubusercontent.com/12507865/128533533-10b68d00-cd24-46bf-94de-b9b9149b7e39.png)

I think it's possible to somehow combine the text props added by Org-indent with Org-transclusion. I just haven't focused on solving this problem. Perhaps you could work out a good way...

**Comment by nobiot** [comment]
This is done in functions `org-transclusion-propertize-transclusion` and `org-transclusion-propertize-source`.
I have a feeling that it might be possible to detect org-indent-mode, grab the relevant text properties, and then add it back combined with what org-transclusion does.

**Comment by randomwangran** [comment]
Hi @nobiot 

I've studied the code and your suggestion.

I see you add the following code to `org-transclusion-remove`

https://github.com/nobiot/org-transclusion/blob/32a699f62597cb74025470d7247d57347e75b354/org-transclusion.el#L524

For my part, my problem still remains. What I found is that I
just need to call `org-indent-add-properties` with the correct
marker position after `org-transclusion-add-all` is executed.

So, I just add

```elisp
(progn
  (org-transclusion-add)
  (org-indent-add-properties (marker-position marker)
                             (marker-position marker)))
```

Instead of your original code:

https://github.com/nobiot/org-transclusion/blob/32a699f62597cb74025470d7247d57347e75b354/org-transclusion.el#L483

I test the modified code on my environment and it is good for me.

But, as you can see, the arguments to `org-indent-add-properties` is not perfect.

I have not to figure out the `end` point after the transcluded
content inserted in the buffer.

**Comment by randomwangran** [comment]
By the way, based on your code at 32a699f62597cb74025470d7247d57347e75b354

If I use this version of `org-transclusion-add-all`, there's no issue with `org-indent-mode`.

```elisp
(defun org-transclusion-add-all (&optional narrowed)
  "Add all active transclusions in the current buffer.

By default, this function temporarily widens the narrowed region
you are in and works on the entire buffer.  Note that this
behavior is important for `org-transclusion-after-save-buffer' in
order to clear the underlying file of all the transcluded text.

For interactive use, you can pass NARROWED with using
`universal-argument' (\\[universal-argument]) to get this
function to work only on the narrowed region you are in, leaving
the rest of the buffer unchanged."
  (interactive "P")
  (save-restriction
    (let ((marker (move-marker (make-marker) (point))))
      (unless narrowed (widen))
      (goto-char (point-min))
      (let ((regexp "^[ \t]*#\\+TRANSCLUDE:"))
        (while (re-search-forward regexp nil t)
          ;; Don't transclude if within a transclusion to avoid infinite
          ;; recursion
          (unless (or (org-transclusion-within-transclusion-p)
                      (plist-get (org-transclusion-keyword-string-to-plist)
                                 :disable-auto))
            (progn
              (org-transclusion-add)
              (org-indent-add-properties (marker-position marker)
                                         (marker-position marker))))))
      (goto-char marker)
      (move-marker marker nil) ; point nowhere for GC
      t)))
```

**Comment by nobiot** [comment]
@randomwangran Thanks for all the comments.

Changing the text-props on the Org-transclusion side is not sufficient and results in incorrect text-properties for headlines, inlinetask, etc.

Try this branch [fix/indent2](https://github.com/nobiot/org-transclusion/tree/fix/indent2).
You then need to override the built-in function as below.
It works on my end for transclusions; however, it does not take care of the source buffer.

```
    (defun org-indent-set-line-properties (level indentation &optional heading)
      "Set prefix properties on current line an move to next one.

    LEVEL is the current level of heading.  INDENTATION is the
    expected indentation when wrapping line.

    When optional argument HEADING is non-nil, assume line is at
    a heading.  Moreover, if it is `inlinetask', the first star will
    have `org-warning' face."
      (let* ((line (aref (pcase heading
			   (`nil org-indent--text-line-prefixes)
			   (`inlinetask org-indent--inlinetask-line-prefixes)
			   (_ org-indent--heading-line-prefixes))
			 level))
	     (wrap
	      (org-add-props
		  (concat line
			  (if heading (concat (make-string level ?*) " ")
			    (make-string indentation ?\s)))
		  nil 'face 'org-indent)))

	(when (org-transclusion-within-transclusion-p)
	  (setq line
		(concat line
			(propertize
			 "x"
			 'display
			 '(left-fringe org-transclusion-fringe-bitmap
				       org-transclusion-fringe))))
	   (setq wrap
		 (concat wrap
			 (propertize
			  "x"
			  'display
			  '(left-fringe org-transclusion-fringe-bitmap
					org-transclusion-fringe)))))

	;; Add properties down to the next line to indent empty lines.
	(add-text-properties (line-beginning-position) (line-beginning-position 2)
			     `(line-prefix ,line wrap-prefix ,wrap)))
      (forward-line))
```

**Comment by randomwangran** [comment]
b3e660d5d4964f3f760403b56e22b80faed4ad6e also works for me. 

Really appreciate it you dig through it and make it compatible with `org-indent-mode`.

Thank you @nobiot .

**Comment by nobiot** [comment]
I’ll keep this issue open. Need to figure out how to incorporate the fix in the main branch; probably as an extension with a switch.

**Comment by nobiot** [comment]
The [`fix/indent2`](https://github.com/nobiot/org-transclusion/tree/fix/indent2) branch adds support for org-indent-mode as an extension. See customizing `org-transclusion-set-extensions`.  `org-transclusion-indent-mode` is off by default.
![image](https://user-images.githubusercontent.com/12507865/131218890-f052fe1f-f4df-458b-aed6-3123f1a64129.png)

If you use Customize, then it will also "require" the extension feature.
If you use `setq` to turn on the extension, you would need to evaluate `(org-transclusion-load-extensions-maybe 'force)` or re-launch Emacs.

**Comment by nobiot** [comment]
@randomwangran I have added this "fix" as a new feature via the extension (README updated). It's now merged with the main at ecf8907. Let me know if you have any questions...

**Comment by nobiot** [comment]
For now, I'm closing this issue with the merge of the feature into main. Let me know if you have any issues.
